src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -215,10 +215,10 @@ internal Binder GetSpeculativeBinder(int position, ExpressionSyntax expression,
             }
 
             // May be binding an expression in a context that doesn't have a LocalScopeBinder in the chain.
-            return new LocalScopeBinder(binder);
+            return new ExpressionVariableBinder(expression, binder);
```

Comment:
At the moment, there is no way to test for an observable difference. However, once we add "match" expression, or an alternative, that would introduce a scope within a portion of an expression, creating ExpressionVariableBinder for the top level scope wouldn't be sufficient. ExecutableCodeBinder, however, will be able to deal with this situation.